scripts/test/Muon/threading_manager_test.py

```diff
@@ -0,0 +1,303 @@
+        # 2 executions per thread
+        num_threads = 4
+        manager = WorkerManager(test_function, num_threads,
+                                lst=[1, 2, 3, 4, 5, 6, 7, 8])
+
+        manager.start()
+        self.Runner(manager)
+
+        self.assertEqual(len(manager.results['results']), 8)
+        for item in [1, 2, 3, 4, 5, 6, 7, 8]:
+            index = [i for i, arg in enumerate(manager.results['lst']) if arg == item][0]
+            self.assertEqual(manager.results['results'][index], 2 * item)
+
+    def test_manager_passes_exceptions_from_evaluation_to_callback(self):
+        manager_finished = mock.Mock()
+        manager_exception = mock.Mock()
+        test_function = lambda lst: test_function_raises_on_n(lst, raise_on=1)
+        manager = WorkerManager(test_function, 4,
+                                callback_on_threads_complete=manager_finished,
+                                callback_on_thread_exception=manager_exception,
+                                lst=[1, 2, 3, 4, 5, 6, 7, 8])
+
+        manager.start()
+        self.Runner(manager)
+
+        self.assertIn(1, manager_exception.call_args[1]['inputs']['lst'])
+
+        self.assertEqual(manager_finished.call_count, 1)
+        self.assertEqual(manager_exception.call_count, 1)
+
+    def test_manager_stores_failed_evaluations(self):
+        manager_finished = mock.Mock()
+        manager_exception = mock.Mock()
+        test_function = lambda lst: test_function_raises_on_n(lst, raise_on=1)
+        manager = WorkerManager(test_function, 4,
+                                callback_on_threads_complete=manager_finished,
+                                callback_on_thread_exception=manager_exception,
+                                lst=[1, 2, 3, 4, 5, 6, 7, 8])
+
+        manager.start()
+        self.Runner(manager)
+
+        self.assertEqual(manager._failed_results['lst'], [1, 5])
+
+    def test_that_once_started_manager_cannot_be_started_again_while_threads_are_running(self):
+        num_threads = 4
+        manager = WorkerManager(test_function, num_threads, lst=[1, 2, 3, 4])
+
+        manager.start()
+        manager.is_running = mock.Mock(return_value=True)
+        self.Runner(manager)
+
+        self.assertRaises(RuntimeError, manager.start)
+
+    def test_that_manager_progress_is_ascending_and_goes_from_0_to_100(self):
+        num_threads = 4
+        manager_progress = mock.Mock()
+        manager = WorkerManager(test_function, num_threads,
+                                callback_on_progress_update=manager_progress,
+                                lst=[1, 2, 3, 4])
+
+        manager.start()
+        self.Runner(manager)
+
+        progress_call_args = [i[0][0] for i in manager_progress.call_args_list]
+        self.assertEqual(progress_call_args[0], 0.0)
+        self.assertEqual(progress_call_args[-1], 100.0)
+        self.assertEqual(sorted(progress_call_args), progress_call_args)
```

Comment:
similar to [this](https://github.com/mantidproject/mantid/pull/23297/files#r211585978)